### PR TITLE
Add VS Code Request Manager module

### PR DESCRIPTION
**Pull request by blrobl**
#### Summary 
The VS Code Request Manager module is used to embed requests in the Web Client (Page Inspector, Extension Management for now) to the AL extension in VS Code to navigate to an object's definition, initiate a debugging session or get the source code of an extension from Git. 

#### Work Item(s) 
Fixes [AB#536537](https://dynamicssmb2.visualstudio.com/1fcb79e7-ab07-432a-a3c6-6cf5a88ba4a5/_workitems/edit/536537)

### CODE REVIEW

**Review by SBalslev — DISMISSED** [review]

File: src/System Application/App/Vs Code Request Management/src/VSCodeRequestManagement.Codeunit.al [hunk 1]
```diff
@@ -0,0 +1,147 @@
+﻿namespace System.Integration;
+
+using System.Apps;
+using System.Reflection;
+using System.Utilities;
+
+codeunit 8033 "VS Code Request Management"
```
Comment (grobyns) [inline src/System Application/App/Vs Code Request Management/src/VSCodeRequestManagement.Codeunit.al]: Do we need VS in there? Do we need Management?

Could it be be called something like Code Requests ? or VSCode Requests? or Source Code Requests?

something along those lines

File: src/System Application/App/Vs Code Request Management/src/VSCodeRequestManagement.Codeunit.al [hunk 2]
```diff
@@ -0,0 +1,147 @@
...
+        AllObjWithCaption: Record AllObjWithCaption;
+        UriBuilder: Codeunit "Uri Builder";
+        VSCodeRequestHelper: DotNet VSCodeRequestHelper;
+        AlExtensionUriTxt: Label 'vscode://ms-dynamics-smb.al', Locked = true;
+        BaseApplicationIdTxt: Label '437dbf0e-84ff-417a-965d-ed2bb9650972', Locked = true;
+        SystemApplicationIdTxt: Label '63ca2fa4-4f03-4f2b-a480-172fef340d3f', Locked = true;
+        ApplicationIdTxt: Label 'c1335042-3002-4257-bf8a-75c898ccb1b8', Locked = true;
+
+    [Scope('OnPrem')]
+    procedure GetUrlToOpenExtensionSource(RepositoryUrl: Text[250]; CommitId: Text): Text
```
Comment (grobyns) [inline src/System Application/App/Vs Code Request Management/src/VSCodeRequestManagement.Codeunit.al]: this does not follow the  facade pattern we use in modules. Please make the Impl. internal and add a public facade with documentation

File: src/System Application/App/Vs Code Request Management/src/VSCodeRequestManagement.Codeunit.al (hunk 1 above)
Comment (grobyns) [reply src/System Application/App/Vs Code Request Management/src/VSCodeRequestManagement.Codeunit.al]: VSCode Integration?

File: src/System Application/App/Extension Management/src/ExtensionManagement.Page.al [hunk 3]
```diff
@@ -182,6 +183,19 @@ page 2500 "Extension Management"
...
+                {
+                    Caption = 'Open Source in VS Code';
+                    Enabled = IsSourceSpecificationAvailable;
+                    Image = Download;
+                    Scope = Repeater;
+                    ToolTip = 'Open the source code for the extension based on the source control information.';
+
+                    trigger OnAction()
+                    begin
+                        Hyperlink(VsCodeRequestManagement.GetUrlToOpenExtensionSource(Rec."Source Repository Url", Rec."Source Commit ID"));
```
Comment (grobyns) [inline src/System Application/App/Extension Management/src/ExtensionManagement.Page.al]: I'd make this so the module has a method which will also do the Hyperlink. When i use the module I'd just want to call something like:
SourceCodeIntegration.ViewSourceCode() 
(maybe with a param or two :P or just pass a Rec? )

and then I wouldn't need to know the 'how' of how to open... could be VS Code, could be a GitHub url...

**Review by grobyns — CHANGES_REQUESTED** [review]
please follow the facade/impl. pattern we use for modules

File: src/System Application/App/Vs Code Request Management/src/VSCodeRequestManagement.Codeunit.al [hunk 4]
```diff
@@ -0,0 +1,147 @@
...
+    var
+        Url: Text;
+    begin
+        if Text.StrLen(RepositoryUrl) <> 0 then begin
+            UriBuilder.Init(AlExtensionUriTxt + '/sourceSync');
+            UriBuilder.AddQueryParameter('repoUrl', RepositoryUrl);
+            if Text.StrLen(CommitId) <> 0 then
+                UriBuilder.AddQueryParameter('commitId', CommitId);
+
```
Comment (krupybalu) [inline src/System Application/App/Vs Code Request Management/src/VSCodeRequestManagement.Codeunit.al]:
```suggestion
            UriBuilder.AddQueryParameter('appid', GetAppIdForObject(ObjectType, ObjectId));

```

File: src/System Application/App/VS Code Integration/src/VsCodeIntegration.Codeunit.al [hunk 5]
```diff
@@ -0,0 +1,34 @@
...
+    Access = Public;
+    InherentEntitlements = X;
+    InherentPermissions = X;
+
+    var
+        AllObjWithCaption: Record AllObjWithCaption;
+        VsCodeIntegrationImpl: Codeunit "VS Code Integration Impl.";
+
+    [Scope('OnPrem')]
+    procedure OpenExtensionSource(var PublishedApplication: Record "Published Application")
```
Comment (JesperSchulz) [inline src/System Application/App/VS Code Integration/src/VsCodeIntegration.Codeunit.al]: Given this is the facade, you must document the public functions.

**Review by JesperSchulz — CHANGES_REQUESTED** [review]
Please add XML docs to public procedures. Then I'll approve!

**Review by JesperSchulz — DISMISSED** [review]

File: src/System Application/App/Extension Management/src/ExtensionManagement.Page.al [hunk 6]
```diff
@@ -182,6 +183,19 @@ page 2500 "Extension Management"
...
+                {
+                    Caption = 'Open Source in VS Code';
+                    Enabled = IsSourceSpecificationAvailable;
+                    Image = Download;
+                    Scope = Repeater;
+                    ToolTip = 'Open the source code for the extension based on the source control information.';
+
+                    trigger OnAction()
+                    begin
+                        VsCodeIntegration.OpenExtensionSource(Rec);
```
Comment (JesperSchulz) [inline src/System Application/App/Extension Management/src/ExtensionManagement.Page.al]: The name of the public function is OpenExtensionSourceInVSCode().

**Review by JesperSchulz — APPROVED** [review]

File: src/System Application/App/VS Code Integration/src/VsCodeIntegration.Codeunit.al [hunk 7]
```diff
@@ -0,0 +1,52 @@
+namespace System.Integration;
```
Comment (darjoo) [inline src/System Application/App/VS Code Integration/src/VsCodeIntegration.Codeunit.al:6]: Add copyright at the top. Same for the other files.
You can find it in the other files like `UserSettings.Codeunit.al`

File: src/System Application/App/VS Code Integration/src/VsCodeIntegration.Codeunit.al [hunk 8]
```diff
@@ -0,0 +1,52 @@
...
+using System.Tooling;
+
+/// <summary>
+/// Provides functionality to embed hyperlinks that send requests to VS Code to navigate to an object's definition in source code
+/// and to open the source code of an extension from Git.
+/// </summary>
+codeunit 8034 "VS Code Integration"
+{
+    Access = Public;
+    InherentEntitlements = X;
```
Comment (darjoo) [inline src/System Application/App/VS Code Integration/src/VsCodeIntegration.Codeunit.al:19]: Should this not be controlled by permissions?

File: src/System Application/App/Extension Management/src/ExtensionManagement.Page.al [hunk 9]
```diff
@@ -366,6 +383,7 @@ page 2500 "Extension Management"
         // Determining Record and Styling Configurations
         IsInstalled := ExtensionInstallationImpl.IsInstalledByPackageId(Rec."Package ID");
         IsTenantExtension := Rec."Published As" <> Rec."Published As"::Global;
+        IsSourceSpecificationAvailable := StrLen(Rec."Source Repository Url") > 0;
```
Comment (darjoo) [inline src/System Application/App/Extension Management/src/ExtensionManagement.Page.al:386]: When is this set? Seems like anyone would be able to click the action as long as the `url` is set. Is that how this should be? Not hidden behind admin permissions etc?

**Review by darjoo — APPROVED** [review]

**Review by SBalslev — APPROVED** [review]

**Pull request merged**